psi4/driver/procrouting/diis.py

```diff
@@ -181,33 +232,170 @@ def extrapolate(self, *args):
         if np.all(diagonals > 0):
             diagonals = diagonals ** (- 0.5)
             B = np.einsum("i,ij,j -> ij", diagonals, B, diagonals)
-            coeffs = np.linalg.lstsq(B, rhs, rcond=None)[0][:-1] * diagonals[:-1]
+            return np.linalg.lstsq(B, rhs, rcond=None)[0][:-1] * diagonals[:-1]
         else:
-            coeffs = np.linalg.lstsq(B, rhs, rcond=None)[0][:-1]
+            return np.linalg.lstsq(B, rhs, rcond=None)[0][:-1]
+
+    def adiis_energy(self, x):
+        x = normalize_input(x)
+        return np.dot(self.adiis_linear, x) + np.einsum("i,ij,j->", x, self.adiis_quadratic, x) / 2
+
+    def adiis_gradient(self, x):
+        c = normalize_input(x)
+        dedc = self.adiis_linear + np.einsum("i,ij->j", c, self.adiis_quadratic)
+
+        norm_sq = (x**2).sum()
+        dcdx = np.diag(x) * norm_sq - np.einsum("i,j->ij", x ** 2, x)
+        dcdx *= 2 / norm_sq**2
+
+        return np.einsum("i,ij->j", dedc, dcdx)
+
+    def adiis_coefficients(self):
+        from scipy.optimize import minimize
+        self.adiis_populate()
+        result = minimize(self.adiis_energy, np.ones(len(self.stored_vectors)), method="BFGS",
+                jac = self.adiis_gradient, tol=1e-6, options={"maxiter": 200})
+
+        if not result.success:
+            raise Exception("ADIIS minimization failed. File a bug, and include your entire input and output files.")
+
+        return normalize_input(result.x)
+
+    def adiis_populate(self):
```

Comment:
Docstring might be good to add even though function name is pretty self-explanatory